lxcfs.c

```diff
@@ -67,30 +83,107 @@ static void users_unlock(void)
 	unlock_mutex(&user_count_mutex);
 }
 
+/*
+ * Simple functions and types to preserve and switch mount namespaces.
+ */
+
+/* initial mount namespace */
+static int mnt_ns_init = -1;
+/* private lxcfs mount namespace */
+static int mnt_ns_lxcfs = -1;
+/* Currently active namespace. When initialized always equal to one of
+ * mnt_ns_init or mnt_ns_lxcfs. */
+static int mnt_ns_active = -1;
+
+/* Close all open file descriptors refering to a namespace. */
+static void close_ns()
+{
+	if (mnt_ns_init > 0) {
+		close(mnt_ns_init);
+		mnt_ns_init = -1;
+	}
+	if (mnt_ns_lxcfs > 0) {
+		close(mnt_ns_lxcfs);
+		mnt_ns_lxcfs = -1;
+	}
+	mnt_ns_active = -1;
+}
+
+/* Open /proc/PID/ns/mnt and save open fd to preserve the mount namespace.
+ * Assumes that @pid is the callers pid and hence, the preserved namespace is
+ * the callers current namespace.
+ */
+static bool preserve_ns(int *ns_store, int *ns_active, int pid)
+{
+	int ret;
+	size_t len = 5 /* /proc */ + 21 /* /int_as_str */ + 7 /* /ns/mnt */ + 1 /* \0 */;
+	char path[len];
+
+	ret = snprintf(path, len, "/proc/%d/ns/mnt", pid);
+	if (ret < 0 || (size_t)ret >= len)
+		return false;
+
+	*ns_store = open(path, O_RDONLY | O_CLOEXEC);
+	if (*ns_store < 0)
+		return false;
+
+	*ns_active = *ns_store;
+
+	return true;
+}
+
+/* Switch caller to namespace identified @ns_switch and set active namespace
+ * @ns_active to the switched namespace @ns_switch. */
+static bool switch_ns(const int ns_switch, int *ns_active)
+{
+	int ret;
+
+	ret = setns(ns_switch, 0);
+	if (ret < 0)
+		return false;
+
+	*ns_active = ns_switch;
+
+	return true;
+}
+
+/*
+ * Functions and types used to reload dynamic library.
+ */
 static volatile sig_atomic_t need_reload;
 
 /* do_reload - reload the dynamic library.  Done under
  * lock and when we know the user_count was 0 */
 static void do_reload(void)
 {
+	if ((mnt_ns_active != -1) && (mnt_ns_active != mnt_ns_init))
```

Comment:
I think that this is a valid case since `do_reload()` is called in `main()` before we have called `preserve_ns()`.